### ISSUE TITLE
False positive unnecessary-dunder-call for __iadd__

### ISSUE THREAD

**Issue by sam-s** [pylint-dev/pylint#8978]
### Bug description

the following code generates an `unnecessary-dunder-call` message
```python
"https"

class Incrementable:
    "class with an incrementable field"
    def __init__(self, a):
        self.a = a
    def __iadd__(self, other):
        if isinstance(other, Incrementable):
            self.a += other.a
            return self
        return NotImplemented

d = {}
d.setdefault("a", Incrementable(6)).__iadd__(Incrementable(7))
```

I see 2 work-arounds, both pollute the namespace:
1. define an intermediate variable:
```
x = d.setdefault("a", Incrementable(6))
x += Incrementable(7)
```

2. Import [`operator`](https://docs.python.org/3/library/operator.html):
```
import operator
operator.iadd(d.setdefault("a", Incrementable(6)), Incrementable(7))
```

Using [`defaultdict`](https://docs.python.org/3/library/collections.html#defaultdict)  is even worse - I will have to convert it back to the ordinary dict for serialization...

### Command used

```shell
pylint z.py
```

### Pylint output

```shell
************* Module z
z.py:14:0: C2801: Unnecessarily calls dunder method __iadd__. Use += operator. (unnecessary-dunder-call)
```

### Expected behavior

no diagnostics - or better one.
since `+=`  requires an extra variable, it is not an acceptable recommendation.

### Pylint version

```shell
pylint 2.17.5
astroid 2.15.6
Python 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0]
```

### OS / Environment

windows, linux

**Comment by nickdrozd** [comment]
Following the suggestion given by the warning results in code that fails to parse:

```python
d.setdefault("a", Incrementable(6)) += Incrementable(7)  # Use += operator
```

```
E0001: Parsing failed: ''function call' is an illegal expression for augmented assignment (<unknown>, line 14)' (syntax-error)
```

**Comment by Pierre-Sassoulas** [comment]
Yes, but I think the message should be modified and still raised (in order to make the code more pythonic like in my proposed snippet). Something like 'create an __add__ method in 'Increment' and use it'

**Comment by sam-s** [comment]
> What about adding the `__add__` function directly in `Incrementable` ?
> 
> ```python
> """https://github.com/pylint-dev/pylint/issues/8978"""
> 
> class Incrementable:
>     """class with an incrementable field"""
>     def __init__(self, a):
>         self.a = a
> 
>     def __add__(self, other):
>         return self.__iadd__(other)
> 
>     def __iadd__(self, other):
>         if isinstance(other, Incrementable):
>             self.a += other.a
>             return self
>         return NotImplemented
> 
> d = {}
> d.setdefault("a", Incrementable(6) + Incrementable(7))
> ```

this does _not_ modify `d["a"]` if it is already present, i.e.

```
d = {"a":Incrementable(2)}
d.setdefault("a", Incrementable(6) + Incrementable(7))
```
is not the same as
```
d = {"a":Incrementable(2)}
d.setdefault("a", Incrementable(6)).__iadd__(Incrementable(7))
```



IOW, this code is _NOT_ equivalent to the code that I posted in all cases.
